backend/app/analysis/crosswords.py

```diff
@@ -0,0 +1,306 @@
+    return ((direction == "across" and column + word_len <= grid_dim[0] - 1)
+            or (direction == "down" and row + word_len <= grid_dim[1] - 1)) and (
+               row >= 0 and column >= 0)
+
+
+def is_cell_free(row, col, grid):
+    """ Checks whether a cell is free.
+    Does not throw if the indices are out of bounds. These cases return as free.
+    """
+    # Negative indices are "legal", but we treat them as out of bounds.
+    if row < 0 or col < 0:
+        return True
+    # marks cell as free
+    try:
+        return grid[row][col] == 0
+    except IndexError:
+        return True
+
+
+def ends_are_isolated(word, row, column, direction, grid):
+    """ Returns whether the given word is isolated (blank before start and after end).
+    """
+    if direction == "across":
+        # If the preceding space isn't empty
+        if not is_cell_free(row, column - 1, grid):
+            return False
+        # If the succeeding space isn't empty
+        if not is_cell_free(row, column + len(word), grid):
+            return False
+    if direction == "down":
+        # If the preceding space isn't empty
+        if not is_cell_free(row - 1, column, grid):
+            return False
+        # If the succeeding space isn't empty
+        if not is_cell_free(row + len(word), column, grid):
+            return False
+
+    return True
+
+
+def collides_with_existing_words(word, row, column, direction, grid):
+    """ Returns whether the given word collides with an existing one.
+    """
+    # iterate over letters and their indices
+    for k, letter in enumerate(list(word)):
+        if direction == "across":
+            # checks for collisions between words
+            if grid[row][column + k] != 0 and grid[row][column + k] != letter:
+                return True
+        if direction == "down":
+            # checks for collisions between words
+            if grid[row + k][column] != 0 and grid[row + k][column] != letter:
+                return True
+
+    return False
+
+
+def add_word_to_grid(grid, word, row, col, direction):
+    """ Adds a possibility to the given grid, which is modified in-place.
+    (see generate_grid)
+    """
+    # Word is left-to-right
+    if direction == "across":
+        grid[row][col:len(list(word)) + col] = list(word)
+
+    # Word is top-to-bottom
+    if direction == "down":
+        # iterates over letters to place letters in correct rows
+        for index, letter in enumerate(list(word)):
+            grid[row + index][col] = letter
+
+
+def make_crossword(grid, word_list, grid_size):
+    """
+    Returns a grid with all the words in a crossword format and their corresponding
+    clues.
+    :param: grid (list): 2D list representing the format of the crossword
+            word_list (list): contains words to be placed into crossword
+            grid_size (int): represents dimension of grid (grid is a square)
+    """
+    first_word = word_list[0]
```

Comment:
We sorted the `word_list` by `len(word)` because it is easier to place the hardest (longest) word into the crossword because it allows us to have more possibilities for connecting words in the future. I have changed the code so that we perform the sort within `make_crossword`.